solr/core/src/java/org/apache/solr/handler/admin/HealthCheckHandler.java

```diff
@@ -88,15 +98,45 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
       return;
     }
 
-    // Set status to true if this node is in live_nodes
-    if (clusterState.getLiveNodes().contains(cores.getZkController().getNodeName())) {
-      rsp.add(STATUS, OK);
-    } else {
+    // Fail if not in live_nodes
+    if (!clusterState.getLiveNodes().contains(cores.getZkController().getNodeName())) {
       rsp.add(STATUS, FAILURE);
       rsp.setException(new SolrException(SolrException.ErrorCode.SERVICE_UNAVAILABLE, "Host Unavailable: Not in live nodes as per zk"));
+      return;
     }
 
-    rsp.setHttpCaching(false);
+    // Optionally require that all cores on this node are active if param 'requireHealthyCores=true'
+    if (req.getParams().getBool(PARAM_REQUIRE_HEALTHY_CORES, false)) {
+      Collection<CloudDescriptor> coreDescriptors = cores.getCores().stream()
+          .map(c -> c.getCoreDescriptor().getCloudDescriptor()).collect(Collectors.toList());
+      List<String> unhealthyCores = findUnhealthyCores(coreDescriptors, clusterState);
+      if (unhealthyCores.size() > 0) {
+          rsp.add(STATUS, FAILURE);
+          rsp.setException(new SolrException(SolrException.ErrorCode.SERVICE_UNAVAILABLE,
+                  "Replica(s) " + unhealthyCores + " are currently initializing or recovering"));
+          return;
+      }
+      rsp.add("message", "All cores are healthy");
+    }
+
+    // All lights green, report healthy
+    rsp.add(STATUS, OK);
+  }
+
+  /**
+   * Find replicas DOWN or RECOVERING, or replicas in clusterstate that do not exist on local node.
+   * We first find local cores which are either not registered or unhealthy, and check each of these against
+   * the clusterstate, and return a list of unhealthy replicas that are part of an active shard for an existing collection
+   * @param cores list of core descriptors to iterate
+   * @param clusterState clusterstate from ZK
+   * @return list of core names that are either DOWN ore RECOVERING on 'nodeName'
+   */
+  static List<String> findUnhealthyCores(Collection<CloudDescriptor> cores, ClusterState clusterState) {
```

Comment:
@shalinmangar Please have a look at the new logic. If all local cores are active and registered, then we do not consult clusterstate at all. And we only consult clusterstate now to filter out replicas from inactive shards.

The only thing I'm unsure of now is whether I feed the correct **slice ID** to `.getActiveSlicesMap().containsKey(**HERE**)`. Is this slice ID the same as `c.getShardId()` on CloudDescriptor? My unit test is a mock, so I cannot be sure :)

Reply:
Looks good to me. Yes, the `c.getShardId()` will give the slice ID so checking that against the active slice map is fine. One gotcha is that the `c.getCoreNodeName()` can return null if a core is not fully loaded yet so you might want to return a count from this method instead of a list of core node names where some of them may be null depending on the current state of the system.

Reply:
Ok. Since I want a more informative error message, I instead did a null-check on `coreNodeName` and use `<collectionName>_<shardId>` as a fallback.

Reply:
I think it might be helpful to name the fallback `unknown:<collectionName>_<shardId>` or `code-loading:<collectionName>_<shardId>` to distinguish from the rest of the list which contains coreNodeNames. Otherwise people might be confused why a core that they can't find in their list is registering as unhealthy.

Reply:
Also this isn't a blocker, but I was wondering if you might have some insight @shalinmangar. When the core directory no longer exists (say for example the disk was wiped before starting the node), the clusterState will register the missing replica(s) as DOWN and the logs will error saying that those cores cannot be found. Will those missing cores still be returned within the `cores.getCores()` call? If so then this is a non-issue, but if they aren't included as I suspect, then this will return healthy even when there are replicas in the clusterState scheduled on the node that are not healthy.

Maybe this just requires a fix in a different part of solr to auto-delete replicas that have cores that are missing on startup. Solr used to auto-recreate the cores, but I think that functionality was removed in Solr 6 or 7.

Reply:
> I think it might be helpful to name the fallback `unknown:<collectionName>_<shardId>` or `code-loading:<collectionName>_<shardId>` to distinguish...

I think I have changed my mind and agree with shalin that it is enough to return a count, and if that count > 0 include in the error msg "N out of M cores are still not healthy". That will avoid the confusion and give a clear and short state to caller. Imagine a node with 3000 cores just having been started but not yet recovered, that list of RECOVERING cores would be huge :) If you really need to know which cores are unhealthy, there are ways to find that elsewhere.

Reply:
> When the core directory no longer exists (say for example the disk was wiped before starting the node), the clusterState will register the missing replica(s) as DOWN and the logs will error saying that those cores cannot be found. Will those missing cores still be returned within the cores.getCores() call?

No, if the core directories themselves have been wiped off then the node cannot return that core in `cores.getCores()`. 

> but if they aren't included as I suspect, then this will return healthy even when there are replicas in the clusterState scheduled on the node that are not healthy.

Yes but that should be okay? The node itself is in fact healthy.

> Maybe this just requires a fix in a different part of solr to auto-delete replicas that have cores that are missing on startup.

Is this a common case? i.e. wiping disks and putting the nodes back in rotation? It is more common to have nodes which have cores that are not in the cluster state and those are unloaded automatically if those nodes come back up. Also, there's node lost trigger which can be used to delete replicas from cluster state for nodes that go away for a long time.

Reply:
So I changed logic to report number of unhealthy cores instead of list of names.
I agree that the corner case of wiping disk is not within scope of this issue.